dali/operators/generic/slice/slice_base.cc

```diff
@@ -64,27 +27,105 @@ DALI_SCHEMA(SliceBase)
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const workspace_t<CPUBackend> &ws) override;
+  void RunImpl(workspace_t<CPUBackend> &ws) override;
+
+  std::vector<SliceArgs>& Args() { return args_; }
+
+ private:
+  std::vector<SliceArgs> args_;
+  kernels::KernelManager kmgr_;
+};
+
+template <typename OutputType, typename InputType, int Dims>
+bool SliceBaseCpu<OutputType, InputType, Dims>::SetupImpl(std::vector<OutputDesc> &output_desc,
+                                                          const workspace_t<CPUBackend> &ws) {
+  const auto &input = ws.template InputRef<CPUBackend>(0);
+  auto in_shape = input.shape();
+  int nsamples = in_shape.num_samples();
+  auto nthreads = ws.GetThreadPool().size();
+  assert(nsamples == static_cast<int>(args_.size()));
+  output_desc.resize(1);
+  output_desc[0].type = TypeInfo::Create<OutputType>();
+  output_desc[0].shape.resize(nsamples, Dims);
+
+  kmgr_.Resize<Kernel>(nthreads, nsamples);
+  kernels::KernelContext ctx;
+  auto in_view = view<const InputType, Dims>(input);
+  for (int i = 0; i < nsamples; i++) {
+    auto in_view = view<const InputType, Dims>(input[i]);
+    auto req = kmgr_.Setup<Kernel>(i, ctx, in_view, args_[i]);
+    output_desc[0].shape.set_tensor_shape(i, req.output_shapes[0][0].shape);
+  }
+  return true;
+}
+
+template <typename OutputType, typename InputType, int Dims>
+void SliceBaseCpu<OutputType, InputType, Dims>::RunImpl(workspace_t<CPUBackend> &ws) {
+  const auto &input = ws.template InputRef<CPUBackend>(0);
+  auto &output = ws.template OutputRef<CPUBackend>(0);
+  int nsamples = input.size();
+
+  auto& thread_pool = ws.GetThreadPool();
+  for (int i = 0; i < nsamples; i++) {
+    thread_pool.DoWorkWithID(
+      [this, &input, &output, i](int thread_id) {
+        auto in_view = view<const InputType, Dims>(input[i]);
+        auto out_view = view<OutputType, Dims>(output[i]);
+        kernels::KernelContext ctx;
+        kmgr_.Run<Kernel>(thread_id, i, ctx, out_view, in_view, args_[i]);
+      });
+  }
+  thread_pool.WaitForWork();
+  output.SetLayout(input.GetLayout());
+}
+
+template <>
+bool SliceBase<CPUBackend>::SetupImpl(std::vector<OutputDesc> &output_desc,
+                                      const workspace_t<CPUBackend> &ws) {
+  const auto &input = ws.template InputRef<CPUBackend>(0);
+  auto input_type = input.type().id();
+  auto ndim = input.shape().sample_dim();
+
+  if (!impl_ || input_type_ != input_type || ndim != ndim_) {
+    impl_.reset();
+    input_type_ = input_type;
+    ndim_ = ndim;
+  }
+  auto output_type = output_type_ == DALI_NO_TYPE ? input_type_ : output_type_;
+
+  VALUE_SWITCH(ndim_, Dims, SLICE_DIMS, (
+    TYPE_SWITCH(input_type_, type2id, InputType, SLICE_TYPES, (
+      if (input_type_ == output_type) {
+        using Impl = SliceBaseCpu<InputType, InputType, Dims>;
+        if (!impl_)
+          impl_ = std::make_unique<Impl>();
+        FillArgs(reinterpret_cast<Impl*>(impl_.get())->Args(), ws);
```

Comment:
A little ugly but IMO better than duplicating the logic to FillArgs (which has nothing to do with CPU/GPU)